presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/PlanCheckerProviderManager.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.sanity;
+
+import com.facebook.presto.spi.plan.PlanCheckerProvider;
+import com.facebook.presto.spi.plan.PlanCheckerProviderFactory;
+import com.facebook.presto.spi.plan.SimplePlanFragmentSerde;
+import com.facebook.presto.sql.planner.PlanFragmenter;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class PlanCheckerProviderManager
+{
+    private final SimplePlanFragmentSerde simplePlanFragmentSerde;
+    private final Map<String, PlanCheckerProviderFactory> providerFactories = new HashMap<>();
```

Comment:
Our other discussion made me realize that this and the below need to use ConcurrentHashMap/CopyOnWriteArrayList.

Reply:
Ok, but from what I understand the plugins are all installed when `PrestoServer` calls `PluginManager.installPlugins()`, is there a way that it's done from multiple threads?

Reply:
I don't think it's ever updated from multiple threads, though we do read from it from multiple threads, so we care about visibility of the writes. And we don't make any guarantees about the writes so it's good practice to be safe.

Reply:
Sounds good, thanks for clarifying